Wayland support (`wlr_gamma_control_unstable_v1`)

minus7 opened the pull request:
This continues the work from #568, replacing `gamma_control` with the newer `wlr_gamma_control_unstable_v1`. This newer protocol changes how gamma ramps are passed to support some graphics card + monitor combinations that have large gamma ramps. The old protocol was limited to 4096 bytes for gamma ramps which the new one avoids by passing ramps via a file descriptor. A snippet of MIT-licensed [code from Weston](https://github.com/wayland-project/weston/blob/master/shared/os-compatibility.c) has been added for that.

I'd like to get this merged upstream to not have it drag on longer. I can squash the changes if you want.

wlroots-based compositors already support this.

minus7 commented:
Looks like the backported pkg.m4 I reverted breaks the build on Ubuntu 14.04. Meh.

zetorian commented:
I can confirm this is working on sway 1.0 with a dual monitor setup, using both internal intel graphics on i915 and nvidia 970m with nouveau, while using prime. great job guys!

colemickens commented:
I'm using this successfully as well, on an XPS 9370 with Intel integration i915.

I do have one small bit of feedback -- Would it be possible to sniff sway/wlroots and default to using the `wayland` method of adjusting gamma if it knows it is running under Wayland?

minus7 commented:
@colemickens I tried to make the autodetect work for Wayland but at least when running Wayland on one tty and X11 on another it doesn't work on X11 anymore

colemickens commented:
@minus7, just realized my own package updated and I didn't notice. Turns out your detection change works well for me (but I also don't run X11 and Wayland simultaneously. Presumably that's a rare scenario.) Thanks for this PR, would be great to see it merged! 👍

xantoz commented:
I've tried this on latest sway and wlroots master (swaywm/sway@cf6edaf26aa1461b89552727d2435d6fe9d0adf3 and swaywm/wlroots@cf6edaf26aa1461b89552727d2435d6fe9d0adf3) and I noticed the following problems:

1) One-shot mode doesn't work. It just flashes briefly.
2) Tearing when setting the gamma ramps. This is relatively visible during the fade-in and fade-out. Maybe this is more of a problem in wlroots, or a limitation in DRM itself, however.

emersion commented:
>One-shot mode doesn't work. It just flashes briefly.

Yes. This is by design. Instead, the oneshot mode should be replaced by a constant mode. I've heard MacOS has the same issue.

>Tearing when setting the gamma ramps. This is relatively visible during the fade-in and fade-out. Maybe this is more of a problem in wlroots, or a limitation in DRM itself, however.

No idea what could cause this. I can't reproduce. Maybe related to your graphics driver.

minus7 committed:
Implement Wayland support

Implement the wlr_gamma_control_unstable_v1 protocol.
The Wayland backend is prioritized over randr since it will fail in
an X11 context anyway (but the randr backend would not fail in a Wayland
context with xwayland support).
The gamma setting is reset when the client disconnects, thus the
behavior for oneshot mode blocks, like for Quartz.

Co-Authored-By: Giulio Camuffo <giuliocamuffo@gmail.com>
Co-Authored-By: Thomas Weißschuh <thomas@t-8ch.de>
Co-Authored-By: Marek Otahal <markotahal@gmail.com>
Co-Authored-By: Dominique Martinet <asmadeus@codewreck.org>

minus7 commented:
I "rebased and squashed" this. That makes it look a lot saner. Honestly, it was a nightmare of branches and merges. I had to resort to `git diff $(git merge-base upstream/master HEAD)..HEAD` and applying that to the upstream master (which oddly enough worked flawlessly then).
I tried attributing everyone that contributed code and mentioning the important changes in the commit message.

minus7 commented:
Would be cool to see this merged, @jonls. Seems like a few people already use this branch through the [AUR](https://aur.archlinux.org/packages/redshift-wlr-gamma-control-git/).

@xantoz: The oneshot issue has been addressed by @martinetd by making it behave like on macOS/Quartz and simply blocking. There's a better solution under discussion in #690

atheriel commented:
What is the relationship between this PR, #704, and #568? They seem largely identical. It would probably be a lot easier for both users and the maintainer if there was an authoritative branch to pull from/review.

emersion commented:
TBH at this point it might be better to establish a proper fork.

atheriel commented:
Only if the maintainers of that fork commit resources to keeping it up to date with the upstream. Otherwise we'll just end up with a further proliferation of these working-at-some-point branches and PRs. (Although I commend @minus7 for recognizing all the previous authors for their work in this particular patch.)

The wlroots project seems like the obvious place for such a fork, if you have the bandwidth for it.

emersion commented:
Considering that there were ~15 commits during last year, I don't think keeping up with upstream is really an issue. @jonls just seems to be lacking time to review PRs. Unfortunately there's no other maintainer, so this project is kind of stuck.

minus7 commented:
> What is the relationship between this PR, #704, and #568? They seem largely identical. It would probably be a lot easier for both users and the maintainer if there was an authoritative branch to pull from/review.

Oh, I didn't know about #704; someone made the effort of squashing the changes there already.
#568 uses an older protocol (see this PR's description) that's no longer used by wlroots-based compositors at least. I don't really know who uses which Wayland gamma interface.

I'm using Redshift on sway, so I'll keep maintaining my branch.

X0rg commented:
(To follow up my comment in #704)

I use [redshift-wlr-gamma-control-git](https://aur.archlinux.org/packages/redshift-wlr-gamma-control-git/) package, which uses [minus7's wayland branch](https://github.com/minus7/redshift/tree/wayland), and it works like a charm with AMDGPU driver on Sway now.

@e00E Does it work for you too with minus7's wayland branch?
Here is the list of versions I use:
sway 1.0-8
wlroots 0.5.0-1
redshift-wlr-gamma-control-git 1.12.r25.geecbfed-1

minus7 commented:
Works since Linux 5.1 (I noticed since I downgraded to 5.0 for a moment)

X0rg commented:
I'm using Linux 5.1.4 with Mesa 19.0.5. Yes, it looks like it was a bug in Linux 5.0.
Thank you very much for your work @minus7!

freed00m commented:
> I use [redshift-wlr-gamma-control-git](https://aur.archlinux.org/packages/redshift-wlr-gamma-control-git/) package, which uses [minus7's wayland branch](https://github.com/minus7/redshift/tree/wayland), and it works like a charm with AMDGPU driver on Sway now.
@X0rg  How? The AUR package fails to build, didn't touch the pkgbuild or the source.
```
./configure: 15410: ./configure: Syntax error: word unexpected (expecting ")")
```

X0rg commented:
@freed00m: IDK, it works with `makepkg`for me.
```
$ echo $PWD && LC_ALL=C makepkg -sif                                                                                                                                   git:(master)
/tmp/redshift-wlr-gamma-control-git
==> Making package: redshift-wlr-gamma-control-git 1.12.r25.geecbfed-1 (Fri Jul  5 20:11:13 2019)
==> Checking runtime dependencies...
==> Checking buildtime dependencies...
==> Retrieving sources...
  -> Updating redshift git repo...
Fetching origin
==> Validating source files with md5sums...
    redshift ... Skipped
==> Extracting sources...
  -> Creating working copy of redshift git repo...
Reset branch 'makepkg'
==> Starting pkgver()...
==> Removing existing $pkgdir/ directory...
==> Starting build()...
Copying file po/Makefile.in.in
autoreconf: Entering directory `.'
autoreconf: running: intltoolize --automake --copy --force
autoreconf: running: aclocal --force -I m4
autoreconf: configure.ac: tracing
autoreconf: running: libtoolize --copy --force
libtoolize: putting auxiliary files in '.'.
libtoolize: copying file './ltmain.sh'
libtoolize: putting macros in AC_CONFIG_MACRO_DIRS, 'm4'.
libtoolize: copying file 'm4/libtool.m4'
libtoolize: copying file 'm4/ltoptions.m4'
libtoolize: copying file 'm4/ltsugar.m4'
libtoolize: copying file 'm4/ltversion.m4'
libtoolize: copying file 'm4/lt~obsolete.m4'
autoreconf: running: /usr/bin/autoconf --force
autoreconf: running: /usr/bin/autoheader --force
autoreconf: running: automake --add-missing --copy --force-missing
configure.ac:56: warning: The 'AM_PROG_MKDIR_P' macro is deprecated, and its use is discouraged.
configure.ac:56: You should use the Autoconf-provided 'AC_PROG_MKDIR_P' macro instead,
configure.ac:56: and use '$(MKDIR_P)' instead of '$(mkdir_p)'in your Makefile.am files.
configure.ac:14: installing './compile'
configure.ac:9: installing './missing'
src/Makefile.am: installing './depcomp'
autoreconf: Leaving directory `.'
checking for a BSD-compatible install... /usr/bin/install -c
[...]
```

minus7 commented:
There's still two things bothering me with this patch:
- After having monitors on standby for a while (over night works well), redshift fails. Still gotta figure out what's wrong there
- Sometimes a single monitor flickers. I guess a gamma update is spread over two frames there.

freed00m commented:
@X0rg 

Thx, I really must find what is the missing dep or what is causing my setup to fail.

This is my log.

[redshift-wlr-gamma-control-git-1.12.r25.geecbfed-1-x86_64-build.log](https://github.com/jonls/redshift/files/3398124/redshift-wlr-gamma-control-git-1.12.r25.geecbfed-1-x86_64-build.log)

X0rg commented:
@freed00m: I have found this: https://github.com/bitlbee/bitlbee-facebook/issues/49#issuecomment-150812762
Did you install the `base-devel` group? It is required to use `makepkg`.

freed00m commented:
@X0rg huray I had only pkgconf missing from base-dev, should a AUR dep :) or am I wrong?

X0rg commented:
@freed00m https://wiki.archlinux.org/index.php/Arch_User_Repository#Getting_started

CameronNemo commented:
@minus7 I would be happy to close out my PR and refer here, but can you pull in the Travis changes so that CI passes?

minus7 commented:
Picked up the travis changes from your branch, @CameronNemo.

Unfortunately now all CI fails.
AppVeyor probably due to Windows being Windows (and removing localtime_r from libc? Unrelated to any changes here anyway).
Travis due to make distcheck failing (happens on your branch too when I try locally). I didn't manage to figure out what's wrong with autotools sorcery there:
`make[3]: *** No rule to make target 'gamma-control-protocol.c', needed by 'all'.  Stop.`

CameronNemo commented:
@minus7 actually that has been started https://github.com/kakurasan/redshift/tree/wayland-meson-testing ! I would prefer if there were a clean commit on top of jonls (autohell). It is still the *only* recognized upstream for distributions.

t-8ch commented:
@minus7 https://github.com/minus7/redshift/pull/2 will fix `make distcheck`

minus7 commented:
Gamma ramp fd leak fixed; calling munmap with the wrong size was the problem.

prometheanfire commented:
Any reason we (gentoo) shouldn't use this patch?  the forked repo's commit (minus the gitignore hunk) is working for me in testing.

Ypnose commented:
Any progress on this ? It is kinda inconvenient for maintainers and users to switch to an external repo for Wayland support.

lheckemann commented:
This doesn't seem to take newly added monitors into account — I have redshift running with `-O 3000` and just plugged in an external monitor and said monitor has a very cold white now. Killing and rerunning redshift fixes it.

zikaeroh commented:
I'm not sure if this is the right place to ask this, but is it expected that `redshift` remains open after doing something like `redshift -b 0.8`? I'm cursed with an OLED screen, so my only source of backlight control comes from gamma changes, and the fact that `redshift` doesn't exit means I can't do something like binding keys to changing the brightness level. Additionally, it crashes if another `redshift` process runs at the same time.

For a good while (essentially the past year) I've just kept a workspace open where I manually go and exit/rerun `redshift` if I'm unhappy with the brightness level, but I figured I'd ask if this is really the expected behavior with this patch.

CameronNemo commented:
@zikaeroh in a way, yes this is the expected behavior. the wlroots protocol requires a process to hold the gamma value steady, it is not a set-and-forget deal. The only good workflow I can think of for you is to store your desired value in a file and read it when starting redshift (via a wrapper script), then bind a second script that rewrites that file and restarts redshift (via the wrapper script) when the brightness keys are pressed. Good luck.

zikaeroh commented:
That's unfortunate, I imagine I'll see a flash of uncorrected frames while that's happening. Maybe I'll have to come up with something that operates over the wlroots gamma protocol directly just for this purpose and forgo redshifting. 🙁 

Is this how the other modes of redshift work, or this patch mean that clients which assume it exits after changing would then hang? Thinking of something like redshift-gtk or similar.

CameronNemo commented:
@zikaeroh other redshift *backends* allow the type of behavior you describe (e.g. the drm or xorg backends). They are not usable on wayland.

martinetd commented:
@zikaeroh see #690 for the behaviour you want (as yes that will flicker if you restart redshift with different values)
I never took the time to properly merge the patches properly but I use them and it works, unfortunately the whole issue needs attention from the maintainer as that really is a design decision here, so I doubt it will move much more on its own... It might make sense to open a PR on minus' branch if enough people need this, but it's also useful to keep things separate for maintainability and let people pick what they need.

zikaeroh commented:
That is awesome, thank you so much for linking that! I now have a fully working backlight setup for the first time in over a year. :smile: 

Did a quick modification of the AUR PKGBUILD and a script adapted from my previous xrandr solution, netting: https://gist.github.com/zikaeroh/b687938443a59d3f0645ed7dabe67aea

TheChymera commented:
I tried to use the latest commit from the respective fork: https://github.com/minus7/redshift/commit/7da875d34854a6a34612d5ce4bd8718c32bec804

With this install log (which looks ok to me) [redshift-wlr.txt](https://github.com/jonls/redshift/files/4327439/redshift-wlr.txt), I get the following output, with no change to the color or brightness of my display:

```
chymera@silenthost ~ $ redshift -v
Temperatures: 6500K at day, 3500K at night
Brightness: 1.00:0.40
Gamma (Daytime): 0.800, 0.800, 0.800
Gamma (Night): 0.800, 0.800, 0.800
Color temperature: 6500K
Brightness: 1.00
Status: Enabled
Period: Night
Color temperature: 3500K
Brightness: 0.40
```

Any idea what's wrong?

(running sway)

marienz commented:
> I tried to use the latest commit from the respective fork: [minus7@7da875d](https://github.com/minus7/redshift/commit/7da875d34854a6a34612d5ce4bd8718c32bec804)
> 
> With this install log (which looks ok to me) [redshift-wlr.txt](https://github.com/jonls/redshift/files/4327439/redshift-wlr.txt),

Build looks ok (gamma-wl.c is built).

> I get the following output, with no change to the color or brightness of my display:
> 
> ```
> chymera@silenthost ~ $ redshift -v
> Temperatures: 6500K at day, 3500K at night
> Brightness: 1.00:0.40
> Gamma (Daytime): 0.800, 0.800, 0.800
> Gamma (Night): 0.800, 0.800, 0.800
> Color temperature: 6500K
> Brightness: 1.00
> Status: Enabled
> Period: Night
> Color temperature: 3500K
> Brightness: 0.40
> ```
> 
> Any idea what's wrong?
> 
> (running sway)

Any chance you have `adjustment-method = randr` in ~/.config/redshift.conf? Try removing any adjustment-method from your redshift config file.

Comparing the code to your output, redshift should be printing the adjustment method it selected after the "Gamma" lines in your output, unless you're in print mode (which it doesn't look like you can be unless you passed -p), or the adjustment method was explicitly selected using the config file or commandline. Testing under Sway 1.4 with `redshift -m randr -v`, I get output very similar to yours and no gamma adjustment. Without `-m randr`, I get "Using method `wayland'." and it works.

TheChymera commented:
@marienz thank you, that worked :) Just for general info, redshift-wlr is now available for Gentoo via the Chymeric Overlay: https://github.com/TheChymera/overlay/tree/master/x11-misc/redshift-wlr

maximbaz commented:
Moving https://github.com/jonls/redshift/pull/704#issuecomment-609973414 here as this is the most recent Wayland PR and I reproduce the issue in this branch too...

-----------

I'm using redshift from this PR on two laptops with Arch Linux, sway and generally identical dotfiles, and on one of them I have weird problems:

- Sometimes screen flickers, for a second colors become broken like on the photo below, then redshift restarts and everything goes back to normal
- If I turn redshift off, colors will get broken like on the image below (running redshift again fixes the colors)

![photo](https://user-images.githubusercontent.com/1177900/78594089-f7957600-7847-11ea-9e74-d31fb2ca17b9.jpeg)

If I run redshift from terminal, I get no errors in the console, but normally I use redshift from a systemd service, and my journalctl is full of the errors below:

```
Process 25051 (redshift) of user 1000 dumped core.

Stack trace of thread 25051:
#0  0x00005631eafb12d7 n/a (redshift + 0xa2d7)
#1  0x00005631eafac2ea n/a (redshift + 0x52ea)
#2  0x00007f4a0904e023 __libc_start_main (libc.so.6 + 0x27023)
#3  0x00005631eafaca6e n/a (redshift + 0x5a6e)
```

I think when my screen flickers for a second, it is exactly the case when redshift coredumps and restarts.

Any hints or ideas on what could be wrong, how to investigate?

emersion commented:
>If I turn redshift off, colors will get broken like on the image below

Try wlroots & Sway master. A fix got merged about gamma ramps recently.

>Any hints or ideas on what could be wrong, how to investigate?

Compile from source with debugging symbols, reproduce the crash, run `coredumpctl gdb` then `bt full`. You'll get a nice backtrace and can start looking into fixing the crash.

clarfon commented:
This has been open since 2018, and it seems like this repo hasn't been maintained much since then. Does anyone have the time/resources to maintain a fork? Or is the maintainer just slow? Not really sure what's up.

CameronNemo commented:
@clarfon I may have some cycles available. See the master branch here https://github.com/CameronNemo/redshift

asadoll commented:
> @clarfon I may have some cycles available. See the master branch here https://github.com/CameronNemo/redshift

Works as expected with both sway and wayfire master branches.

vperilla commented:
Tested with [Hikari](https://hikari.acmelabs.space/) and works fine

TheChymera commented:
@CameronNemo is there no chance to get this merged?

martinetd commented:
> @CameronNemo is there no chance to get this merged?

This isn't CameronNemo's repo here. He made fork where it is merged (and redshift got renamed to gammastep after removing windows/mac support, with readme pointing to a gitlab repo that doesn't exist and no way to create issues...), but the only one who can make changes here is @jonls when he can find time to get around it.
He did get around to review the dbus PR recently so all hope isn't lost yet, maybe :)

ubone commented:
I built redshift from minus7's repo and the only issue I have is that -PO does not quit as usual.

CameronNemo commented:
If I cleaned up my repo and marked a release, would anyone be willing to package it for:

* Arch (AUR)
* Nix
* Debian
* Gentoo
* anything else

?

TheChymera commented:
> If I cleaned up my repo and marked a release, would anyone be willing to package it for:
> 
>     * Gentoo
>
> ?

Would this be a “permanent” fork? I.e. do you have the interest and time availability to merge future work by the original upstream, or continue to develop it your own way? I would much prefer if we could somehow convince the principal (@jonls ?) upstream to merge this so that development efforts are consolidated in one place. Forks are great when upstream becomes unresponsive, but they're annoying to maintain in parallel and have a bad habit of running out of steam and becoming outdated once the original issue was satisfied...

CameronNemo commented:
True. my biggest concern would be missing out on the dbus service work that is going on in another fork (https://github.com/dkondor/redshift/tree/dbus_new).

But I cannot wait any longer. I will fork by the end of the month and mark a release. I would hope some distros pick it up, because I know many people want wayland support, but that happening is not a blocker to me forking.

lheckemann commented:
@CameronNemo yes, I'm happy to take care of the nix package.

c4rlo commented:
For Arch Linux, I've just submitted @CameronNemo's [fork](https://github.com/CameronNemo/redshift) (_gammastep_) to the AUR as [gammastep-git](https://aur.archlinux.org/packages/gammastep-git).

CameronNemo commented:
What a coincidence. I just put some finishing touches on the branch, made the gitlab repository public, and tagged a release. I did most of it last night but wanted to sleep on it.

c4rlo commented:
Sweet! I've pushed it to the AUR as [gammastep](https://aur.archlinux.org/packages/gammastep), and have removed gammastep-git from the AUR.

CameronNemo commented on src/os-compatibility.c:
```diff
@@ -0,0 +1,148 @@
+int create_tmpfile_cloexec(char *tmpname) {
+	int fd;
+	mode_t prev_umask = umask(0066);
+#ifdef HAVE_MKOSTEMP
```
Will this ever be defined? There is no AC_CHECK_FUNCS([mkostemp]) in the configure.ac file.

https://gitlab.com/chinstrap/gammastep/-/commit/47502deb942dff764d9f896f1004eaaa9c3cf4ef

CameronNemo commented on src/os-compatibility.c:
```diff
@@ -0,0 +1,148 @@
+		return -1;
+	}
+
+#ifdef WLR_HAS_POSIX_FALLOCATE
```
This should be HAVE_POSIX_FALLOCATE, and see above

CameronNemo commented on src/gamma-wl.c:
```diff
@@ -0,0 +1,372 @@
+			return -1;
+		}
+
+		void *ptr = mmap(NULL, total_bytes,
```
if gamma_size is 0 (i.e. when the output does not support setting gamma), the total_bytes will be zero and mmap will error with EINVAL.

https://gitlab.com/chinstrap/gammastep/-/commit/fdfec2bfb4ee8a44ae503a7c63a073453a17f176

CameronNemo reviewed (changes requested):
Needed to apply a couple of commits on top of this.

nbraud commented:
FYI, I [wrote a patch] that updates redshift's AppArmor profile to make it work with this PR.

[wrote a patch]: https://salsa.debian.org/debian/redshift/-/blob/c70977fea0a55bf14dbbd87ba8ac783ad73c796a/debian/patches/Fix-AppArmor-profile-for-Wayland.patch

rustycl0ck commented:
ping @jonls , would be great to have this reviewed/merged

archseer commented:
@rustycl0ck We've all migrated over to the https://gitlab.com/chinstrap/gammastep fork

minus7 commented:
Another alternative (with less features) is [wlsunset|https://sr.ht/~kennylevinsen/wlsunset/]. I'll close this MR since I don't use nor maintain it anymore.
